pkg/skaffold/cluster/minikube.go

```diff
@@ -0,0 +1,204 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"os/exec"
+
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
+	k8s "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/client"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/context"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+var GetClient = getClient
+
+// To override during tests
+var (
+	mkBinaryFunc  = minikubeBinary
+	k8sConfigFunc = context.GetRestClientConfig
+)
+
+type Client interface {
+	// IsMinikube returns true if the given kubeContext maps to a minikube cluster
+	IsMinikube(kubeContext string) (bool, error)
+	// MinikubeExec returns the Cmd struct to execute minikube with given arguments
+	MinikubeExec(arg ...string) (*exec.Cmd, error)
+}
+
+type clientImpl struct{}
+
+func getClient() Client {
+	return clientImpl{}
+}
+
+func (clientImpl) IsMinikube(kubeContext string) (bool, error) {
+	// short circuit if context is 'minikube'
+	if kubeContext == constants.DefaultMinikubeContext {
+		return true, nil
+	}
+	_, err := mkBinaryFunc()
+	if err != nil {
+		return false, nil // minikube binary not found
+	}
+
+	if ok, err := matchNodeLabel(kubeContext); err != nil {
+		return false, nil
+	} else if ok {
+		logrus.Debugf("Detected minikube cluster for context %s due to matched labels", kubeContext)
+		return true, nil
+	}
+
+	if ok, err := matchProfileAndServerURL(kubeContext); err != nil {
+		return false, nil
+	} else if ok {
+		logrus.Debugf("Detected minikube cluster for context %s due to matched profile name or server url", kubeContext)
+		return true, nil
+	}
+	return false, nil
+}
+
+func (clientImpl) MinikubeExec(arg ...string) (*exec.Cmd, error) {
+	return minikubeExec(arg...)
+}
+
+func minikubeExec(arg ...string) (*exec.Cmd, error) {
+	b, err := mkBinaryFunc()
+	if err != nil {
+		return nil, fmt.Errorf("getting minikube executable: %w", err)
+	}
+	return exec.Command(b, arg...), nil
+}
+
+func minikubeBinary() (string, error) {
+	execName := "minikube"
+	if found, _ := util.DetectWSL(); found {
+		execName = "minikube.exe"
+	}
+	filename, err := exec.LookPath(execName)
+	if err != nil {
+		return "", errors.New("unable to find minikube executable. Please add it to PATH environment variable")
+	}
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		return "", fmt.Errorf("unable to find minikube executable. File not found %s", filename)
+	}
+	return filename, nil
+}
+
+func matchNodeLabel(profileName string) (bool, error) {
+	client, err := k8s.Client()
+	if err != nil {
+		return false, fmt.Errorf("getting Kubernetes client: %w", err)
```

Comment:
can we just return false or true here and in `matchProfileAndServerURL`. 

That ways, we can simply check
```
def IsMinikube (kubeContext string ) bool {
    .... 
   return matchNodeLabel || matchProfileAndServerURL 
}
```

Reply:
I'll end up having a bunch of log messages in these functions. I think I'll leave it as is and log only in the top method